Sources/NIOPosix/SelectableEventLoop.swift

```diff
@@ -421,16 +430,169 @@ Further information:
+                if #available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *) {
+                    erasedUnownedJob.unownedJob.runSynchronously(on: self.asUnownedSerialExecutor())
+                } else {
+                    fatalError("Tried to run an UnownedJob without runtime support")
+                }
+            #endif
+            }
+        }
+    }
+    
+    private func runLoop() -> NIODeadline? {
+        // We need to ensure we process all tasks, even if a task added another task again
+        while true {
+            // TODO: Better locking
+            let nextReadyDeadline = self._tasksLock.withLock { () -> NIODeadline? in
+                // We expect empty self.taskCopy, to put a new batch of tasks into
+                assert(self.tasksCopy.isEmpty)
+
+                var moreImmediateTasksToConsider = !self._immediateTasks.isEmpty
+                var moreScheduledTasksToConsider = !self._scheduledTasks.isEmpty
+
+                guard moreImmediateTasksToConsider || moreScheduledTasksToConsider else {
+                    // We will not continue to loop here. We need to be woken if a new task is enqueued.
+                    self._pendingTaskPop = false
+                    
+                    // Reset nextReadyDeadline to nil which means we will do a blocking select.
+                    return nil
+                }
+
+                // We only fetch the time one time as this may be expensive and is generally good enough as if we miss anything we will just do a non-blocking select again anyway.
+                let now: NIODeadline = .now()
+                var nextScheduledTaskDeadline = now
+
+                while self.tasksCopy.count < Self.taskCopyBatchSize &&
+                        (moreImmediateTasksToConsider || moreScheduledTasksToConsider) {
+                    // We pick one item from self._immediateTasks & self._scheduledTask per iteration of the loop.
+                    // This prevents one task queue starving the other.
+                    if moreImmediateTasksToConsider, let task = self._immediateTasks.popFirst() {
+                        self.tasksCopy.append(task)
+                    } else {
+                        moreImmediateTasksToConsider = false
+                    }
+
+                    if moreScheduledTasksToConsider, let task = self._scheduledTasks.peek() {
+                        if task.readyTime.readyIn(now) <= .nanoseconds(0) {
+                            self._scheduledTasks.pop()
+                            self.tasksCopy.append(.function(task.task))
+                        } else {
+                            nextScheduledTaskDeadline = task.readyTime
+                            moreScheduledTasksToConsider = false
+                        }
+                    } else {
+                        moreScheduledTasksToConsider = false
+                    }
+                }
+
+                assert(self.tasksCopy.count <= Self.taskCopyBatchSize)
```

Comment:
This assertion seems like it is straightforward to break. We can add two tasks in each loop iterator above, but we only check the length once. I think the loop above needs an extra length check.

Reply:
Good catch. Reorganising the loop to check the length of the `taskCopy` array at each point we're mutating (adding to) it.

Reply:
Heh, I tried to write a test for this, and realised the code before was fine, just. It's because:
* if `_scheduledTasks` and `_immediateTasks` are non-empty, we'd be adding 2 items to the `tasksCopy` each time round the loop. And `taskCopyBatchSize` is even, so one test every 2 items is sufficient
* if either `_scheduledTasks` or `_immediateTasks` run out of tasks, we'd be testing the size of `tasksCopy` once per each addition. This applies even if we just run out when we're terminating the batch.

This wasn't intentional micro-optimisation, so I'm adding the additional length check back to make the code safer against future refactoring (including making the size of the batch odd).

Reply:
Yeah my concern was very much about future refactors :wink: